src/cMakeTaskProvider.ts

```diff
@@ -0,0 +1,147 @@
+      command: "build"
+    };
+    const task = new vscode.Task(definition, vscode.TaskScope.Workspace, taskName, CMakeTaskProvider.CMakeSourceStr);
+    task.group = vscode.TaskGroup.Build;
+    task.detail = "CMake template build task";
+    result.push(task);
+    return result;
+  }
+
+  public resolveTask(task: CMakeTask): CMakeTask | undefined {
+    const execution: any = task.execution;
+    if (!execution) {
+        const definition: CMakeTaskDefinition = <any>task.definition;
+        const scope: vscode.WorkspaceFolder | vscode.TaskScope = vscode.TaskScope.Workspace;
+        const resolvedTask: CMakeTask = new vscode.Task(definition, scope, definition.label, CMakeTaskProvider.CMakeSourceStr,
+            new vscode.CustomExecution(async (resolvedDefinition: vscode.TaskDefinition): Promise<vscode.Pseudoterminal> =>
+              new CustomBuildTaskTerminal(resolvedDefinition.command, this.defaultTarget, resolvedDefinition.options, this.cmakeDriver)
+            ), []); // TODO: add problem matcher
+        return resolvedTask;
+    }
+    return undefined;
+  }
+
+}
+
+class CustomBuildTaskTerminal implements vscode.Pseudoterminal , proc.OutputConsumer {
+  private writeEmitter = new vscode.EventEmitter<string>();
+  private closeEmitter = new vscode.EventEmitter<number>();
+  public get onDidWrite(): vscode.Event<string> { return this.writeEmitter.event; }
+  public get onDidClose(): vscode.Event<number> { return this.closeEmitter.event; }
+  private endOfLine: string = "\r\n";
+
+  constructor(private command: string, private defaultTarget: string, private options?: { cwd?: string }, private cmakeDriver?: CMakeDriver) {
+  }
+
+  output(line: string): void {
+    this.writeEmitter.fire(line + this.endOfLine);
+  }
+
+  error(error: string): void {
+    this.writeEmitter.fire(error + this.endOfLine);
+  }
+
+  async open(_initialDimensions: vscode.TerminalDimensions | undefined): Promise<void> {
+      // At this point we can start using the terminal.
+      this.writeEmitter.fire(localize("starting.build", "Starting build...") + this.endOfLine);
+      await this.doBuild();
+  }
+
+  close(): void {
+      // The terminal has been closed. Shutdown the build.
+  }
+
+  private async doBuild(): Promise<any> {
+    if (this.command !== "build") {
+      this.writeEmitter.fire(localize("not.a.build.command", "\"{0}\" is not a recognized build command.", this.command) + this.endOfLine);
+      this.closeEmitter.fire(-1);
+      return;
+    }
+    let buildCommand: proc.BuildCommand | null;
+    let cmakePath: string = "CMake.EXE";
+    let args: string[] = [];
+
+    if (this.cmakeDriver) {
+      buildCommand = await this.cmakeDriver.getCMakeBuildCommand(this.defaultTarget);
+      if (buildCommand) {
+        cmakePath = buildCommand.command;
+        args = buildCommand.args ? buildCommand.args : [];
+      }
+    }
+    if (this.options?.cwd) {
+      this.options.cwd = util.resolveVariables(this.options.cwd);
```

Comment:
>util.resolveVariables [](http://example.com/codeflow?start=25&length=21)

I see you copied this over from cpptools.  cmake-tools has it's own resolver.  Look at `expand.expandStringHelper`.  To supply the `opts`, do a FindAllReferences and see what the other functions are doing.  There may be a way to make this better in the future (e.g. pass in a driver or extension object and we'll fill it out for you), but that would be for a separate PR.

Reply:
in future we can use `expandStringHelper(str, this.cmakeDriver.expansionOptions)` to resolve variables other than `cwd`, 
currently vscode resolves `cwd `in `resolveTask`.